client/src/controllers/RulesController.ts

```diff
@@ -0,0 +1,134 @@
+import { Controller } from '@hotwired/stimulus';
+
+import { castArray } from '../utils/castArray';
+import { debounce } from '../utils/debounce';
+
+/**
+ * Adds the ability for a controlled form element to conditionally
+ * enable targeted elements based on the data from the controlled form
+ * along with a set of rules to match against that data.
+ *
+ * @example - Enable a button if a specific value is chosen
+ * ```html
+ * <form data-controller="w-rules" data-action="change->w-rules#resolve">
+ *   <select name="fav-drink" required>
+ *     <option value="">Select a drink</option>
+ *     <option value="coffee">Coffee</option>
+ *     <option value="other">Other</option>
+ *   </select>
+ *   <button type="button" data-w-rules-target="enable" data-w-rules='{"fav-drink": ["coffee"]}'>
+ *     Continue
+ *   </button>
+ * </form>
+ * ```
+ */
+export class RulesController extends Controller<HTMLFormElement> {
+  static targets = ['enable'];
+
+  /** Targets will be enabled if the target's rule matches the scoped form data, otherwise will be disabled. */
+  declare readonly enableTargets: HTMLElement[];
+  /** Value set on connect to ensure there are valid targets available, to avoid running rules when not needed. */
+  declare readonly hasEnableTarget: boolean;
+
+  declare active: boolean;
+  declare rulesCache: Record<string, [string, string[]][]>;
+
+  initialize() {
+    this.rulesCache = {};
+    this.resolve = debounce(this.resolve.bind(this), 50);
+  }
+
+  connect() {
+    this.checkTargets();
+  }
+
+  /**
+   * Checks for any targets that will mean that the controller needs to be active.
+   */
+  checkTargets() {
+    this.active = this.hasEnableTarget;
+  }
+
+  /**
+   * Resolve the conditional targets based on the form data and the target(s)
+   * rule attributes and the controlled element's form data.
+   */
+  resolve() {
+    if (!this.active) return;
+
+    const formData = new FormData(this.element);
+
+    this.enableTargets.forEach((target) => {
+      const shouldEnable = this.findRules(target).every(
+        ([fieldName, validValues]) => {
+          // reminder that forms can have multiple values for the same field name
+          const allFieldValues = formData.getAll(fieldName);
+          // reminder that checkbox fields will NOT appear in FormData unless checked
+          const fieldValues = allFieldValues.length > 0 ? allFieldValues : [''];
+          return validValues.some((validValue) =>
```

Comment:
Worth thinking about: this currently matches "any" valid value in the rules. I can see this controller being useful for enforcing "all" the valid values, e.g. requiring a set of fields to be selected in a select multiple widget. There's no such need for it in the codebase currently, but I assume it would be as simple as changing the logic from `.some()` to `.every()` here.

If that's the case, that makes me wonder: should we maybe rename the `data-w-rules` attribute to something like `data-w-rules-any` to match any value, with the possibility of adding `data-w-rules-all` support later? (Naming TBD, could also use `data-w-rules-some` and `data-w-rules-every` if that's preferred. I think `any` and `all` ala Python is more readable though)

I think even if we don't add support for "all" in this PR, `data-w-rules-any` makes it clear that you only expect at an "any" match with the values in the rules, whereas with `data-w-rules` it's not very obvious. The docstring of the controller at the moment also does not make it clear that it's an "any" matching.

Reply:
Yeah good thinking, I'd started down that track (something similar, I'll have to find my older branches) but felt it was too much 'future use'. I think maybe we'll start with a sensible default behaviour and see what use cases come up.

I also plan to enhance to support enable and disable options and was going to use the data attribute to support this. Such as `data-w-rules-enable`, `data-w-rules-disable`, `data-w-rules-show` and `data-w-rules-hide`.

I'll have a think about this and see what could work, I like the idea of the default being that all must match. It feels more intuitive.

Just brainstorming, but we could pass in a special name to the rule to support this. For example, `noindex` or an empty name are not valid input name saccording to the spec. https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#naming-form-controls:-the-name-attribute

We could use this to support something like `data-w-rules='{"":"any",...}'`, `data-w-rules='{"noindex":"any",...}'`.

I was thinking of these directives; `all`, `any`, `non`, `one`. This would probably be abstracted away at the Django widget level with specific kwargs that make sense in python. I've used these names in a different project, three characters and alpha sorting gives a logical default of 'all'.

Reply:
OK, this suggestion is one level lower than my reply is about. As it's not the ruleset that this line relates to but the value matching for a field.

In which case, I actually think we should leave it as is for now. This change is a bit painful to make and in my mind makes it harder for the case where you want to support more flexible matching.

I feel that this PR as is has a solid enough starting point for us to build on, we may want to consider how the value matching works later. We could support inner objects within the value matching, not just an array but maybe an object that allows more complex things.

Thanks @laymonage  though, this is good input, I will do my best to sum this up into a new issue for future enhancements both at the ruleset (per element) level and at the field value matching. I also think we will need to consider complex cases such as streamfield eventually.

Reply:
I have added a reminder about this on the Stimulus migration project - will assess when we have more momentum on known use cases first.

https://github.com/orgs/wagtail/projects/14/views/1?pane=issue&itemId=89429564

Reply:
PR up that includes some of these improvements, without going to deep into 'future' use cases though.
https://github.com/wagtail/wagtail/pull/12667